Add a tool to support generating large amounts of data

revans2 opened the pull request:
I know the patch is rather large.  This provides a lot of functionality, but I will file follow on issues to add in support for the rest of the missing data types and some skewed data distributions.

Suggestions or comment are very welcome.

revans2 commented:
build

gerashegalov commented on integration_tests/src/main/spark311/scala/org/apache/spark/sql/tests/datagen/DataGenExprBase.scala:
```diff
@@ -0,0 +1,28 @@
+{"spark": "311"}
+{"spark": "312"}
+{"spark": "313"}
+spark-rapids-shim-json-lines ***/
+
+package org.apache.spark.sql.tests.datagen
```
The conversion path in `shimplify` inserts `spark-rapids-shim-json-lines` without leaving a blank line over the package declaration. Deleting for consistency:

```suggestion
spark-rapids-shim-json-lines ***/
package org.apache.spark.sql.tests.datagen
```

gerashegalov commented on integration_tests/src/main/spark320/scala/org/apache/spark/sql/tests/datagen/DataGenExprBase.scala:
```diff
@@ -0,0 +1,44 @@
+{"spark": "333"}
+{"spark": "340"}
+{"spark": "341"}
+spark-rapids-shim-json-lines ***/
+
+package org.apache.spark.sql.tests.datagen
```
```suggestion
spark-rapids-shim-json-lines ***/
package org.apache.spark.sql.tests.datagen
```

revans2 commented:
build

abellina commented:
This is mostly a general comment as I haven't gone through the code in detail yet. 

I often find it a friction point in the unit tests to generate dataframes that are meaningful. Could we have a .toBatch in addition to .toDF in this code (this could be follow on work), and should this code be accessible from the unit tests as well? We could port unit tests that use dataframes to this new approach.

revans2 commented:
>  Could we have a .toBatch in addition to .toDF in this code

I assume that you mean you want to create a GpuColumnarBatch from the data? That was one place that I didn't think about adding in more plug-ability.  I think it should be doable, but it does come with a number of issues.

1. I wanted this to be somewhat agnostic of the Rapids Accelerator or any other accelerator.  As such I would want to make sure that it is a separate piece of code that the unit tests would provide to get the generated data out.
2. The parallel worlds class loader would cause similar issues in that I don't think I could get access to GpuColumnBatch directly anyways. I think the solution would be the same as 1
3. We would have to adjust/redo packaging because we don't really want the unit tests to depend on the integration tests jar.  I am happy to move things around, and possibly put this into a totally separate repo, with a different release cycle if we want to.  Just needed access to the shimplify scripts to make it work for multiple different versions of Spark. I could have done it with a UDF, but the API to do it requires me to set a special config to enable a UDF where Spark does not determine the return type.

abellina commented:
It seems `big_data_gen.scala` is mostly generic code that (hopefully) shouldn't change between spark versions, please correct me if I am wrong.

Could this be in its own module, or sql-plugin? Then the classes that are in the integration_tests shim specific directory could still see it via the integration_test pom?

`DataGenExpr` seems to be the only class with spark specifics and it is rather small. We could duplicate it in tests/ and integration_tests if needed?

The other place that seems like a good fit to me is just the sql-plugin. I know it feels a little odd to put it there and is against the first point to keep it agnostic, but unless we replicate the shims it makes a bit harder.

Again, this is a general comment, I think it would be good to get this in and start using it, then someone could take the time to make it usable from the unit tests as well?

revans2 commented:
> Could this be in its own module, or sql-plugin?

own module? Yes sql-plugin? No this is really a stand alone thing. It is tied to Spark, but not the accelerator plugin and I don't want to couple them unless there is a really good reason to do so.

> `DataGenExpr` seems to be the only class with spark specifics and it is rather small. 

Right now it is, but there are a lot of types that are not supported by all of the versions of Spark that we do support so we have to shim them too. I don't want to just hack something together to make a test work. If we want this for multiple different tests, then we should make it stand alone.

There is another issue here too.  The framework is very generic, but it is designed with scale testing in mind.  There are some situations in the future where there could be conflicts between what a unit test wants and what a scale test wants. In those cases the scale testing side is going to win out.  We do have data generation tools in the unit tests, like for CastOpSuites it might be better to look at some of them and adjust those rather than try to make this fit.  Also generally we want to favor integration tests over unit tests, so it would take a lot to convince me that this is the route we want to take.

jlowe commented on integration_tests/src/main/scala/org/apache/spark/sql/tests/datagen/big_data_gen.scala:
```diff
@@ -0,0 +1,1923 @@
+}
+
+object ByteGen {
+  def asByte(a: Any): Byte = a match {
+    case n: Byte => n
+    case n: Short if n <= Byte.MaxValue && n >= Byte.MinValue =>
+      n.toByte
+    case n: Int if n <= Byte.MaxValue && n >= Byte.MinValue =>
+      n.toByte
+    case n: Long if n <= Byte.MaxValue && n >= Byte.MinValue =>
+      n.toByte
+    case other =>
+      throw new IllegalArgumentException(s"a byte value range only supports " +
+          s"byte values found $other")
+  }
```
If the intent is to allow type conversions as long as the original value is preserved in the new form, we could simplify this to:
```suggestion
  def asByte(a: Any): Byte = {
    val converted = a.toByte
    if (converted == a) {
      converted
    } else {
      throw new IllegalArgumentException(s"a byte value range only supports " +
          s"byte values found $other")
    }
  }
```
Which would then allow byte conversions for floating point, decimals, etc as long as their value was completely convertible into a byte (which seems to meet the spirit of what was coded for short, int, and long).

revans2 replied:
`a` is an Any. There is no `toByte` for `a`. a Scala Byte class only inherits from AnyVal, so there is not even a type hierarchy to integer values/etc. I'll try to make it work with java's Number. It looks like scala plays nicely with that.

jlowe commented on integration_tests/src/main/scala/org/apache/spark/sql/tests/datagen/big_data_gen.scala:
```diff
@@ -0,0 +1,1923 @@
+    tables.foreach {
+      case (name, gen) =>
+        val subPath = path + "/" + name
+        gen.toDF(spark, numParts).write.mode("overwrite").parquet(subPath)
```
Do we want this to be dangerous by default?  Wondering if we need to worry about accidental data loss due to overwrite.

razajafri commented on integration_tests/DATA_GEN.md:
```diff
@@ -0,0 +1,402 @@
+
+To get started with big data generation the first thing you need to do is 
+to include the appropriate jar on the classpath for your version of Apache Spark. 
+Note that this does not run on the GPU, it is a standalone, but it does use
```
```suggestion
Note that this does not run on the GPU, it is standalone, but it does use
```

razajafri commented on integration_tests/DATA_GEN.md:
```diff
@@ -0,0 +1,402 @@
+
+Generating values with a desired unique count is great, but we also want
+to be able to generate values with various different distributions of keys.
+By default, the generated values have a `FlastDistribution`, meaning that
```
Did you mean FlatDistribution or UniformDistribution? 
```suggestion
By default, the generated values have a `FlatDistribution`, meaning that
```

razajafri commented on integration_tests/DATA_GEN.md:
```diff
@@ -0,0 +1,402 @@
+### DistinctDistribution
+
+The `DistinctDistribution` generates seeds from min seed to max seed, but
+walks them in a non-sequential order. The data generation functions also maps
```
```suggestion
walks them in a non-sequential order. The data generation function also maps
```

razajafri commented on integration_tests/DATA_GEN.md:
````diff
@@ -0,0 +1,402 @@
+```
+
+Or you could run it with the RAPIDS Accelerator where the data generations is still
+done on the CPU and cut the time down to just 75,696 ms
````
75,696 ms on which hardware?

revans2 replied:
My desktop. It is more of a job at CPUs than anything.

razajafri reviewed (commented):
Just documentation comments so far. Still haven't gone through everything

razajafri commented on integration_tests/DATA_GEN.md:
```diff
@@ -0,0 +1,402 @@
+|  26|-1617384775774370579|
+| -43|-2697761218773508623|
++----+--------------------+
+dataTable.configureKeyGroup(Seq("a", "b"), CorrelatedKeyGroup(1, 0, 1), FlatDistribution())
```
Maybe a hint about what the params to `CorrelatedKeyGroup` mean will avoid a user looking elsewhere for an answer?

revans2 commented on integration_tests/src/main/scala/org/apache/spark/sql/tests/datagen/big_data_gen.scala:
https://docs.scala-lang.org/style/files.html

But It should be lowercase camel not snake case.

razajafri replied:
Thanks, yes, this is a multi-unit file and should be lower case camel case

revans2 commented:
build

razajafri commented on integration_tests/src/main/scala/org/apache/spark/sql/tests/datagen/bigDataGen.scala:
```diff
@@ -0,0 +1,1940 @@
+   * Names of the tables.
+   */
+  def tableNames: Seq[String] = tables.keys.toSeq
+
```
If you think this has value we can do it as a follow-on

revans2 commented:
@razajafri I added a `get` in addition to `apply` to be in keeping with how scala does this. I also added in the API you wanted for converting to multiple data frames and another one that lets you get the tables gens back as a Map so you can do whatever processing on it that you want.

Please take another look

razajafri commented on integration_tests/src/main/scala/org/apache/spark/sql/tests/datagen/bigDataGen.scala:
```diff
@@ -1804,23 +1822,40 @@ class DBGen {
    */
   def apply(name: String): TableGen = tables(name.toLowerCase)
 
+  def get(name: String): Option[TableGen] = tables.get(name)
```
What made you decide to not convert this to a lower case just like you did in `apply`? Trying to know the thought process

revans2 replied:
I forgot. Good catch.

razajafri replied:
Actually, I understand why you didn't convert it to lower case here. I think what I don't understand is why convert it to lowercase in `apply`

revans2 replied:
Spark by default is case insensitive.  A table names "FOO" and "foo" are the same, which is why I normalize them all to be lowercase. I do that when I add the table to the map too.

razajafri commented on integration_tests/src/main/scala/org/apache/spark/sql/tests/datagen/bigDataGen.scala:
```diff
@@ -1804,23 +1822,40 @@ class DBGen {
+
+  /**
+   * Convert all of the tables into dataframes
+   * @param spark the session to use for the conversion
```
nit: empty line need here?

revans2 commented:
build

revans2 commented:
build

wjxiz1992 commented:
Can we add `setLength()` parameter(better all user-configurable APIs) in the README directly with usage example? It's often used when generating customer data according to my previous experience. Otherwise users need to jump to source code to see more.

gerashegalov commented on integration_tests/src/main/scala/org/apache/spark/sql/tests/datagen/bigDataGen.scala:
```diff
@@ -0,0 +1,1978 @@
+  }
+
+  def get(name: String): Option[ColumnGen] = {
+    columns.find {
+      case (childName, _) =>
+        childName.equalsIgnoreCase(name)
+    }.map(_._2)
```
nit:
```suggestion
    columns.collectFirst {
      case (childName, colGen) if childName.equalsIgnoreCase(name) => colGen
    }
```

revans2 commented:
@wjxiz1992 how do I reproduce the failures with JDK 17? The warnings appear to be related to a different version of scala, not java, so I am confused why JDK 17 is changing the version of scala or causing scala to complain.

wjxiz1992 commented:
> @wjxiz1992 how do I reproduce the failures with JDK 17? The warnings appear to be related to a different version of scala, not java, so I am confused why JDK 17 is changing the version of scala or causing scala to complain.

Ignore this error, after I upgrade my maven and use `-Pjdk17`, I can build the jar successfully.

revans2 commented:
build

revans2 commented:
@jlowe @gerashegalov could you both look again. I think I have gotten through all of the review comments and would like to get this in sooner than later.

@wjxiz1992 if you have more comments I am happy to do them, but possibly in a follow on PR.

abellina reviewed (approved):
I'd like to start playing with this, and I think we should get this in and iterate on it if we find issues.

I got through the API high level, and it makes sense. I need to wrap my head on how to use the distributions but I think that comes with use.
